File: solutions/csharp/diffie-hellman/1/DiffieHellman.cs
using System.Numerics;

public static class DiffieHellman
{
    private static readonly Random Random = new();

    public static BigInteger PrivateKey(BigInteger primeP) =>
        Random.Next(2, (int)primeP);

    public static BigInteger PublicKey(BigInteger primeP, BigInteger primeG, BigInteger privateKey) =>
        BigInteger.ModPow(primeG, privateKey, primeP);

    public static BigInteger Secret(BigInteger primeP, BigInteger publicKey, BigInteger privateKey) =>
        BigInteger.ModPow(publicKey, privateKey, primeP);
}

File: solutions/kotlin/diffie-hellman/1/src/main/kotlin/DiffieHellman.kt
import java.math.BigInteger
import java.util.Random

object DiffieHellman {
    private val random = Random()

    fun privateKey(prime: BigInteger): BigInteger = randomNumber(prime - BigInteger.ONE) + BigInteger.ONE

    fun publicKey(p: BigInteger, g: BigInteger, privKey: BigInteger): BigInteger = g.modPow(privKey, p)

    fun secret(prime: BigInteger, publicKey: BigInteger, privateKey: BigInteger): BigInteger =
        publicKey.modPow(privateKey, prime)

    private fun randomNumber(upperBound: BigInteger): BigInteger =
        BigInteger(upperBound.bitLength(), random).mod(upperBound)
}
